profiling/src/internal/observation/observations.rs

```diff
@@ -7,43 +7,58 @@ use super::super::Sample;
 use super::trimmed_observation::{ObservationLength, TrimmedObservation};
 use crate::internal::Timestamp;
 use std::collections::HashMap;
+use crate::collections::identifiable::*;
+use lz4_flex::frame::FrameEncoder;
+use std::io::Write;
 
 struct NonEmptyObservations {
     aggregated_data: HashMap<Sample, TrimmedObservation>,
-    timestamped_data: Vec<TrimmedTimestampedObservation>,
+    compressed_timestamped_data: FrameEncoder<Vec<u8>>,
     obs_len: ObservationLength,
 }
 
-// Timestamp and TrimmedObservation are both 64bit values
-// Using a 32 bit SampleId would still take 64 bits due to padding
-// So just put the Sample in here
-type TrimmedTimestampedObservation = (Sample, Timestamp, TrimmedObservation);
-
 #[derive(Default)]
 pub struct Observations {
     inner: Option<NonEmptyObservations>,
 }
 
 /// Public API
 impl Observations {
-    pub fn add(&mut self, sample: Sample, timestamp: Option<Timestamp>, values: Vec<i64>) {
-        if let Some(inner) = &self.inner {
-            inner.obs_len.assert_eq(values.len());
+    pub fn init_timeline(&mut self, values_size: usize) {
+        if let Some(_inner) = &self.inner {
+            panic!("Should never happen!");
         } else {
+            // // Create buffer, pre-touch all elements to avoid any laziness
+            let mut timestamped_data_buffer: Vec<u8> = vec![0; 1_048_576];
```

Comment:
The values above are in kilobytes; e.g. taking the prototype 223 468k => 223MiB.

I actually did make some experiments to try to account for all the memory, sharing here as well:

Prototype, no timeline: ~218 MiB (e.g. since we pre-initialize the space anyway, and the Ruby profiler keeps two profiles active at all time)

Prototype, timeline, no serialization: ~219 MiB (e.g. the extra space seen above this is memory needed during serialization, since timeline profiles, even compressed, are bigger)